### PR TITLE
🪐 Support for JupyterLab 4.x (take 2!)

### PR DESCRIPTION
**Pull request by agoose77**
@rowanc1 this PR dispenses with the attempt to move everything into a MIME renderer. Following the work on #142, it's clear to me that there was a reasonable amount of trying to fit a square peg into a round hole. 

We should squash the commits when merging, as their lack of meaning represents the haphazard way that I built it out.

In this PR there is still be "double rendering" when executing a Markdown cell with expressions. I think this is probably reasonable; we probably don't want to hang the UI whilst waiting for a long expression to render. Maybe others will have different expectations, in which case we can rethink the approach.

In this rework, we're back to a distinct cell-vs-mime pathway. This seems like a more natural fit, given that fragmented documents need a lot more work than a single-document renderer. From #142, each cell is given a `MySTWidget` that re-uses the React rendering code. Placeholders _should_ now be supported; if a cell has a placeholder during `render()`, only the AST is parsed over a render.

I anticipate that the initial load of a notebook might need optimising; this involves an N^2 render, because all cells are updated. We could detect this with an `initialRender` flag, and defer document-level rendering until the final cell.

### CODE REVIEW

**Comment by fperez** [comment]
For reference, in case anyone wants to quickly test this, updating the command kindly provided by @parmentelat in #142, use:

```
 pip install git+https://github.com/executablebooks/jupyterlab-myst.git@agoose77/feat-jupyterlab-v4-next
```

and restart your JupyterLab session.

If you work in a mamba/conda-managed environment and want to limit that pip only installs this extension, you can use:

```
mamba install webcolors uri-template fqdn isoduration arrow
```

to get the extension's dependencies via mamba, and pip will only add `jupyterlab-myst`.

**Comment by rowanc1** [comment]
I will take a look in the next few days, I know that I was intentionally triggering a render because there was the 1s delay for showing the change on the checkbox, so I was faking it and rendering faster. If that has been fixed through other ways then we might want to simplify the logic of the task list.

Thanks for making progress here, your summary makes a lot of sense, and we were fighting with two very different things going on between the markdown documents and notebooks.

**Comment by agoose77** [comment]
> Thanks for making progress here, your summary makes a lot of sense, and we were fighting with two very different things going on between the markdown documents and notebooks.

I'm aware that you've been of the opinion that we should release to get something out, so I appreciate the patience in seeing this through. I was not feeling happy that #142 is good enough to release, but I am happier about this direction, even if we do have some double rendering still.

**Comment by rowanc1** [comment]
Just very excited to get it out there and support JupyterLab 4! 🚀 (sorry if you felt pressured!)

The approach here looks much more sensible, I will make the tests go green and do my best to fix up the task lists in my testing.

**Comment by rowanc1** [comment]
I spent much of today digging into this but have very little to show. The double rendering is persistent even when I can't see anything that would change. I will continue to debug -- one place I have limited visibility is what the jupyterlab-react interface is doing, and if that is forcing rerenders somehow.

A rerender is fine as it should be pretty efficient, however the flash is not good.

There are a few other bugs reintroduced (images, md render error, tasks disabled) that I might move onto instead to get this closer to a merge.

**Comment by rowanc1** [comment]
@agoose77 I spent another few hours this evening getting into the react loop, I really don't think it is us, I am pretty sure the react node is getting killed by Jupyter, leading to a flash. We can "fix" it by changing the timing on the delayed trigger from the markdown `activityMonitor`. I think changing that from 1000ms --> 100ms makes the double render mostly go away in my testing. What I think is really happening is that the flash is happening **before** you shift-enter, which looks better. This means if you have two _views_ of the notebook open, it is not super great. I am really not sure what Jupyter is doing with the react connection here.

With that fix, the UX seems great. The code, less so. :)

That is my recommendation for now, and then otherwise all other functionality is good to go. Updated libraries and fixed a few small other issues. I also enabled abbreviations and terms/glossaries.

Let's fix the screenshot test after we squash this into main (otherwise we have to force-push, run the update, un-force push).

A mini-bug is that the task list is enabled in the markdown preview, and it doesn't do anything.

**Comment by rowanc1** [comment]
I also pushed a change and now have tested in JupyterLab3. Everything works except for the cell execution, I also don't think that would be hard to make it compatible, but we don't have to do that out of the gate. :)

@agoose77 can you take a look & test and then we can release!

**Review by rowanc1 — APPROVED** [review]
🚀

**Comment by rowanc1** [comment]
Slight improvements on metadata set/get/delete that is backwards compatible. Now everything except inline widgets are working in `jlab3.x`!

My plan is to squash, merge and release a major version in the next hour.

**Comment by parmentelat** [comment]
I have taken the liberty to quote your code in this discourse question here, hope you don't mind

https://discourse.jupyter.org/t/jlab-extensions-that-target-both-jlab3-and-jlab4/20100

I was about to go down a similar route in order to support both jlab 3 and 4, so seeing your code prompted me to ask the question on behalf of all extension writers ;-)

and many thanks btw for releasing 2.0 🥇

**Comment by rowanc1** [comment]
Nice -- looking forward to seeing the responses!

**Comment by parmentelat** [comment]
yeah, me too, insofar as I get any ;-)

File: pyproject.toml [hunk 1]
```diff
@@ -1,53 +1,55 @@
...
     "Programming Language :: Python",
     "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3.7",
     "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
 ]
 dependencies = [
+    "jupyter_server>=2.0.1,<3"
```
Comment (jtpio) [inline pyproject.toml:25]: Do we need `jupyter_server` as a dependency?

Comment (jtpio) [reply pyproject.toml:25]: This seems to be causing installation issue in JupyterLite, as reported in https://github.com/jupyterlite/jupyterlite/issues/1122.

**Pull request merged**